File: commands/general/spongetext.ts
import { ChatInputCommandInteraction, EmbedBuilder, SlashCommandBuilder } from 'discord.js';
import Bot from '../../bot/bot';
import SlashCommand from '../../bot/slashcommand';

export default class SpongeText extends SlashCommand {
  constructor(client: Bot) {
    const data = new SlashCommandBuilder()
      .setName('spongetext')
      .setDescription('Convert text into sPOnGe cASe')
      .addStringOption(option => option
        .setName('text')
        .setDescription('The text to convert')
        .setRequired(true)
      );

    super(data as SlashCommandBuilder, client);
  }

  async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    const input = interaction.options.getString('text')!;
    let output = '';

    const data = Array.from({ length: input.match(/\S/g)!.length }, () => false);

    // First two characters are always different
    data[0] = Math.random() < 0.5;
    data[1] = !data[0];

    for (let i = 2; i < data.length; i++) {
      // No more than two consecutive upper/lower case characters in a row
      if (data[i - 1] === data[i - 2]) {
        data[i] = !data[i - 1];
        continue;
      }
      data[i] = Math.random() < 0.5;
    }

    let j = 0;
    for (let i = 0; i < input.length; i++) {
      if (input[i] === ' ') {
        output += ' ';
        continue;
      }
      output += data[j++] ? input[i].toUpperCase() : input[i].toLowerCase();
      // Increase j only when used
    }

    const embed = new EmbedBuilder()
      .setColor(this.client.config.colors.BLUE)
      .setTitle('Here\'s your converted text')
      .setDescription(output);

    interaction.reply({ embeds: [embed], ephemeral: true });
  }
}

File: src/commands/general/spongetext.rs
use regex::Regex;
use serenity::all::{CommandInteraction, CommandOptionType};
use serenity::async_trait;
use serenity::builder::{CreateCommand, CreateCommandOption, CreateEmbed};
use serenity::prelude::Context;

use crate::color::Colors;
use crate::commands::SlashCommand;
use crate::database::Database;
use crate::interaction::{BetterResponse, InteractionCustomGet};
use crate::Result;

#[derive(Default)]
pub struct SpongeText;

#[async_trait]
impl SlashCommand for SpongeText {
  fn register(&self) -> CreateCommand {
    CreateCommand::new("spongetext")
      .description(format!(
        "Convert text into {}",
        randomize_case("sponge case".to_string())
      ))
      .add_option(CreateCommandOption::new(CommandOptionType::String, "text", "The text to convert").required(true))
  }

  async fn execute(&self, ctx: &Context, interaction: &CommandInteraction, _: &Database) -> Result<()> {
    let input = interaction.get_string("text").unwrap();
    let output = randomize_case(input);

    let embed = CreateEmbed::new()
      .color(Colors::Blue)
      .title("Here's your converted text")
      .description(output);

    interaction.reply_embed_ephemeral(&ctx.http, embed).await?;
    Ok(())
  }
}

fn randomize_case(input: String) -> String {
  let mut output = String::new();

  let re = Regex::new(r"\S").unwrap();
  let length = re.captures_iter(input.as_str()).count();
  let mut is_capitalized = Vec::new();

  // First two cases are random
  is_capitalized.push(rand::random::<bool>());
  is_capitalized.push(rand::random::<bool>());

  for i in 2..length {
    is_capitalized.push(
      // Two consecutive cases at the most, exclude whitespace
      if is_capitalized[i - 1] == is_capitalized[i - 2] {
        !is_capitalized[i - 1]
      } else {
        rand::random::<bool>()
      }
    );
  }

  let mut j: usize = 0;
  for ch in input.chars() {
    if ch == ' ' {
      output.push(' ');
      continue;
    }
    output.push(if is_capitalized[j] {
      ch.to_ascii_uppercase()
    } else {
      ch.to_ascii_lowercase()
    });
    j += 1; // Increase only when used
  }

  output
}
